documentation/Extension_Point_WebScripts.md

```diff
@@ -0,0 +1,306 @@
+
+## Exception handlers
+
+If you want to reuse logic to handle exceptions across WebScript methods or classes, consider annotating a dedicated method with `@ExceptionHandler`:
+
+```java
+@ExceptionHandler(IllegalArgumentException.class)
+protected void handleIllegalArgument(IllegalArgumentException exception, WebScriptResponse response) {
+    response.setStatus(400);
+}
+```
+
+## Before
+
+Any generic initialization can be put in WebScript initialization method, marked with `@Before`. (inherited from superclasses)
+
+## Resolutions
+
+So far, the output of a webscript was either handled manually (using response argument) or by returning a model from the controller method.
+
+In order to improve control flow and enable reuse of output strategies, you can use/implement resolutions.
+
+All webscript Uri methods can return an implementation of `com.github.dynamicextensionsalfresco.webscripts.resolutions.Resolution`.
+
+This is a simple interface, capable of handling a webrequest by itself:
+
+```java
+public interface Resolution {
+    void resolve(final AnnotationWebScriptRequest request, final AnnotationWebscriptResponse response,
+                 final ResolutionParameters params) throws Exception;
+}
+```
+
+Whenever your return a model from your Uri method, you're implicitly returning a `TemplateResolution` from your method.
+
+The power of a Resoluton comes with reuse:
+if your webscript needs to output JSON, you can return an anonymous implementation of `JsonWriterResolution`:
+
+
+```java
+@Uri("/api/hello")
+public Resolution sayHello() {
+  return new JsonWriterResolution() {
+    void writeJson(JSONWriter jsonWriter) throws JSONException {
+      jsonWriter.object()
+        .key("message").value("hello")
+        .endObject();
+    }
+  };
+}
+```
+
+If you want to use another Json serialisation library, create a new subclass of `JsonResolution`.
+
+Because resolutions are returned, they automatically control the flow of your Uri method.
+
+## Request and ResponseBody
+
+Since version 1.7, there is support for the Spring [RequestBody](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestBody.html) and [ResponseBody](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/ResponseBody.html) annotation. These work with using HttpMessageConverters. At the moment, there are only 2 HttpMessageConverterts available by default. The Jackson converter for json, and the JAXB converter for xml.
+
+### RequestBody
+The `@RequestBody` annotation will _resolve the method argument depending on the content type of the request._ When annotating a method parameter with this, the body of the request will be automatically resolved to the parameter object. Dynamic extensions will select a resolver based on the `Content-Type` header from the request. If no resolver for the specified content-type can be found, a `RuntimeException` will be thrown. 
```

Comment:
If I read this for the first time, I would think that I need to define these resolvers myself, but if I POST some json that can deserialize to the MyObject below, I'd be fine, right? This is a very useful feature so maybe we should make a bit clearer how useful this is.

Reply:
I moved this above the `Resolutions` section to avoid confusion & 'bump it's importance'.
I think if you read this before the `Resolutions` section, you'dd be less confused?

Reply:
That would be good, yes.